### ISSUE TITLE
Consider building WebGL on to of gfx-rs

### ISSUE THREAD

**Issue by jdm** [servo/servo#21320]
```
<pcwalton> I know nobody wants to hear this :) but my inclination with Servo's implementation of WebGL is to implement it on top of gfx-rs instead of using the native driver
<pcwalton> rationale: drivers are garbage, Apple is killing GL, and if gfx-rs becomes the unique substrate for WebGL and WebGPU it makes our life easier when remoting it
```
@pcwalton Could you elaborate more on this idea? Are there any clear downsides to that plan?

**Comment by SimonSapin** [comment]
@pcwalton, would this project be similar in size to re-implementing all of ANGLE? (~510k lines of C/C++ counted by Tokei.)

**Comment by pcwalton** [comment]
One interesting possibility would be *using* ANGLE. It seems that Google is
working on a Vulkan backend, and gfx-rs has a Vulkan front-end. So, in
theory, it should be possible to hook ANGLE up to gfx-rs and have a full
gfx-rs stack, without us having to do the work of writing a GL
implementation.

cc @kvark

**Comment by kvark** [comment]
> I know nobody wants to hear this :)

I can't think of a better elevator pitch intro :)

> would this project be similar in size to re-implementing all of ANGLE? (~510k lines of C/C++ counted by Tokei.)

Wow, that's about 15x the size of gfx-rs (not counting dependencies). I surely hope the effort is not going to be that big.

> One interesting possibility would be *using* ANGLE.

It's a neat shortcut, especially given that Google/Angle is heavily investing into Vulkan currently. But ultimately I'd prefer us not depending on Angle in the long term.

> The major question mark, to me, is whether it makes sense to have a gfx-rs remoting layer so we can run everything that runs on top of gfx-rs out of process.

That's indeed an interesting question...

I believe, at this point we should just focus on WebGPU and implement WebGL on top of it. Clearly that's not going to be as fast as implementing WebGL directly, but any performance-hungry applications should move to WebGPU anyway, and WebGL can't survive much longer if the native GL is getting deprecated. Ideally, it's just a WASM compatibility library that runs WebGL on top of WebGPU, and we don't have any special support inside the browsers.

With this strategy, it's not the gfx-rs remoting layer we need to implement but rather WebGPU remoting, used by WebGL as well. As a bonus, WebGPU is higher level than WebGL so should be easier to implement upon.

**Comment by SimonSapin** [comment]
> Ideally, it's just a WASM compatibility library that runs WebGL on top of WebGPU, and we don't have any special support inside the browsers.

Maybe some sites/apps will opt to do that if the benefits outweight the cost of shipping that compat layer across the network, but WebGL is not going away anytime soon from browser implementations. Even if it’s not necessarily the fastest path, we’ll need to keep supporting it for a *long* time.

**Comment by pcwalton** [comment]
There's https://github.com/Think-Silicon/GLOVE which I just saw go by on Twitter as well.

**Comment by kvark** [comment]
@pcwalton that's an interesting project to look deeper at. My first reaction is suspicion due to the lack of notion of OpenGL CTS coverage and zero issues/PRs (isolated one-man show).

**Comment by pcwalton** [comment]
Yeah, I'm skeptical of it too, but it goes to show that an implementation of GLES 2 may not be *that* much work.

**Comment by nox** [comment]
> Apple is killing GL

If (when?) Apple finally kills GL, they will have to make WebKit use something else for their implementation of WebGL. IMO we should just wait and see what they do there.

**Comment by kvark** [comment]
@nox using GL framework is already triggering warnings/errors (suppressible, yet) when building MacOS apps on OSX 10.14 with XCode 10. Next thing you know - they can deny new GL applications to the AppStore. They can still make an exception for their own Safari, but of course this is rather speculative.

**Comment by Protagores** [comment]
> I believe, at this point we should just focus on WebGPU and implement WebGL on top of it. Clearly that's not going to be as fast as implementing WebGL directly, but any performance-hungry applications should move to WebGPU anyway, and WebGL can't survive much longer if the native GL is getting deprecated. Ideally, it's just a WASM compatibility library that runs WebGL on top of WebGPU, and we don't have any special support inside the browsers.

@kvark I'm afraid there's still a lot of "legacy" hardware out there which works fairly well with OpenGL (ES) 2.0/3.0 but might not work with Vulkan (on top of which I suppose WebGPU would normally be implemented) due to lack of support in drivers, as well as general shortcomings/compatibility barriers of their hardware architecture (try ridiculously obscure pipelines, lack of VM, shader engines with aggressively reduced VLIW instruction sets to which arbitrary SPIR-V code doesn't really map well, and so on). I'd prefer to have a fallback option open for such older hardware anyway, otherwise I expect the owners of such hardware to be more incentivized to drop "too demanding" browsers, rather than upgrade their devices.

**Comment by kvark** [comment]
I agree. Fwiw, gfx-rs has backends in development for OpenGL/GLES and D3D11, although clearly going WebGL -> WebGPU -> gfx-rs -> D3D11 is going to be slower than just WebGL -> Angle -> D3D11.

**Comment by kvark** [comment]
Another interesting opportunity here is to implement GLES on top of WebGPU (or, more concretely, https://github.com/gfx-rs/wgpu). It's significantly simpler than GLES on gfx-rs, yet would allow us to avoid Angle entirely. We'd also get remoting for free, because wgpu-rs is designed with it in mind.
